File: src/main/java/com/kbkapps/playgba/cpu/Registers.java
package com.kbkapps.playgba.cpu;

import com.kbkapps.playgba.cpu.constants.Flags;

import static com.kbkapps.playgba.cpu.constants.Flags.*;

public class Registers {
    public static final int SYSTEM_OR_USER = 0;
    public static final int INTERRUPT_REQUEST = 1;
    public static final int SUPERVISOR = 2;
    public static final int ABORT = 3;
    public static final int UNDEFINED = 4;
    static final char CPSR = 'C';
    static final char SPSR = 'S';
    //Registers 0-12
    private final int[] unbankedReg = new int[13];
    //Registers 13(SP) and 14(LR)
    private final int[] bankedReg = new int[10];
    private final int[] savedStatusReg = new int[5];
    public boolean irqEnable;
    public boolean thumbMode;
    private int mode;
    private boolean privileged;
    //PC
    private int reg15 = 4;
    private int currentStatusReg;

    public Registers() {
        System.out.println("Start State: " + this);
    }

    public Registers(int[] unbankedReg, int[] bankedReg, int[] savedStatusReg, int r15, int cpsr) {
        System.arraycopy(unbankedReg, 0, this.unbankedReg, 0, 13);
        System.arraycopy(bankedReg, 0, this.bankedReg, 0, 10);
        System.arraycopy(savedStatusReg, 0, this.savedStatusReg, 0, 5);
        reg15 = r15;
        currentStatusReg = cpsr;
        setControlBits();
    }

    private void setControlBits() {
        irqEnable = (currentStatusReg & 0x80) == 0;
        thumbMode = (currentStatusReg & 0x20) != 0;
        int modeMeta = currentStatusReg & 0x1F;
        switch (modeMeta) {
            case 0:
            case 16:
                mode = SYSTEM_OR_USER;
                privileged = false;
                break;
            case 2:
            case 18:
                mode = INTERRUPT_REQUEST;
                break;
            case 3:
            case 19:
                mode = SUPERVISOR;
                break;
            case 23:
                mode = ABORT;
                break;
            case 27:
                mode = UNDEFINED;
                break;
            case 31:
                mode = SYSTEM_OR_USER;
                privileged = true;
                break;
            default:
                throw new IndexOutOfBoundsException("FIQ:" + modeMeta);
        }
    }

    int getReg(int index) {
        if (index < 0)
            throw new IndexOutOfBoundsException();
        if (index < 13)
            return unbankedReg[index];
        if (index < 15)
            return bankedReg[getBankedIndex(index)];
        if (index == 15)
            return reg15;
        throw new IndexOutOfBoundsException();
    }

    int getPC() {
        return reg15 - getStepAmount();
    }

    private int getBankedIndex(int index) {
        index -= 13;
        return index * 5 + mode;
    }

    void setReg(int index, int data) {
        if (index < 0)
            throw new IndexOutOfBoundsException();
        if (index < 13)
            unbankedReg[index] = data;
        else if (index < 15)
            bankedReg[getBankedIndex(index)] = data;
        else if (index == 15) {
            assert data > 0;
            reg15 = data;
        } else
            throw new IndexOutOfBoundsException();
    }

    void setPC(int data) {
        int newPC = data + getStepAmount();
        assert newPC > 0;
        reg15 = newPC;
    }

    public int getPSR(char type) {
        if (type == CPSR)
            return currentStatusReg;
        if (type == SPSR)
            return savedStatusReg[mode];
        throw new IndexOutOfBoundsException();
    }

    public void setPSR(char type, int data) {
        if (type == CPSR) {
            currentStatusReg = data;
            setControlBits();
        } else if (type == SPSR)
            savedStatusReg[mode] = data;
        else throw new IndexOutOfBoundsException();
    }

    public void setFlags(int flags) {
        currentStatusReg = currentStatusReg & 0xF_FF_FF_FF;
        currentStatusReg = currentStatusReg | flags;
    }

    public boolean canExecute(Flags cond) {
        //NegativeZeroCarryOverflow(V)----  --------  --------  IrqFiqThumbMMMMMode
        if (cond == AL) return true;
        else if (cond == MI) return currentStatusReg < 0;
        else if (cond == PL) return currentStatusReg >= 0;
        else if (cond == EQ) return (currentStatusReg & 0x40_00_00_00) != 0;
        else if (cond == NE) return (currentStatusReg & 0x40_00_00_00) == 0;
        else if (cond == CS) return (currentStatusReg & 0x20_00_00_00) != 0;
        else if (cond == CC) return (currentStatusReg & 0x20_00_00_00) == 0;
        else if (cond == VS) return (currentStatusReg & 0x10_00_00_00) != 0;
        else if (cond == VC) return (currentStatusReg & 0x10_00_00_00) == 0;
        else if (cond == HI) return canExecute(CS) && canExecute(NE);
        else if (cond == LS) return canExecute(CC) || canExecute(EQ);
        else if (cond == GE) return canExecute(MI) == canExecute(VS);
        else if (cond == LT) return canExecute(MI) != canExecute(VS);
        else if (cond == GT) return canExecute(NE) && canExecute(GE);
        else if (cond == LE) return canExecute(EQ) || canExecute(LT);
        else throw new IndexOutOfBoundsException(cond.name());
    }

    public void step() {
        reg15 += getStepAmount();
    }

    private int getStepAmount() {
        return thumbMode ? 2 : 4;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj instanceof Registers)
            return this.toString().equalsIgnoreCase(obj.toString());
        return false;
    }

    @Override
    public String toString() {
        StringBuilder regState = new StringBuilder();
        for (int i = 0; i < 13; i++)
            regState.append("r").append(i).append(":0x").append(Integer.toUnsignedString(unbankedReg[i], 16)).append(" ");
        regState.append("r13:0x").append(Integer.toUnsignedString(getReg(13), 16)).append(" ");
        regState.append("r14:0x").append(Integer.toUnsignedString(getReg(14), 16)).append(" ");
        regState.append("r15:0x").append(Integer.toUnsignedString(reg15, 16)).append(" ");
        regState.append("\nsp: ");
        for (int i = 0; i < 5; i++)
            regState.append("0x").append(Integer.toUnsignedString(bankedReg[i], 16)).append(" ");
        regState.append("\nlr: ");
        for (int i = 5; i < 10; i++)
            regState.append("0x").append(Integer.toUnsignedString(bankedReg[i], 16)).append(" ");
        regState.append("\nspsr: ");
        for (int i = 0; i < 5; i++)
            regState.append("0x").append(Integer.toUnsignedString(savedStatusReg[i], 16)).append(" ");
        regState.append("\nmode:").append(Integer.toUnsignedString(mode, 16)).append(" ")
                .append("privileged:").append(privileged).append(" ")
                .append("pc:0x").append(Integer.toUnsignedString(getPC(), 16)).append(" ")
                .append("cpsr:0x").append(Integer.toUnsignedString(currentStatusReg, 16));

        return regState.toString();
    }

    public String getMode() {
        switch (mode) {
            case SYSTEM_OR_USER:
                if (privileged)
                    return "sys";
                else
                    return "usr";
            case INTERRUPT_REQUEST:
                return "irq";
            case SUPERVISOR:
                return "svc";
            case ABORT:
                return "abt";
            case UNDEFINED:
                return "und";
            default:
                return null;
        }
    }

    public void setMode(String mode) {
        switch (mode) {
            case "sys":
                this.mode = SYSTEM_OR_USER;
                privileged = true;
                break;
            case "usr":
                this.mode = SYSTEM_OR_USER;
                privileged = false;
                break;
            case "irq":
                this.mode = INTERRUPT_REQUEST;
                break;
            case "svc":
                this.mode = SUPERVISOR;
                break;
            case "abt":
                this.mode = ABORT;
                break;
            case "und":
                this.mode = UNDEFINED;
                break;
            default:
                throw new IndexOutOfBoundsException(mode);
        }
    }
}

File: Registers.cpp
#include <iostream>
#include "FailureCodes.h"
#include "CommonDS.h"
#include "Registers.h"

using namespace std;

Registers::Registers(/* args */)
{
}

Registers::~Registers()
{
}

void Registers::setControlBits() {
    irqEnable = (currentStatusReg & 0x80) == 0;
    int modeMeta = currentStatusReg & 0x1F;
    switch (modeMeta) {
        case 0:
        case 16:
            mode = SYSTEM_OR_USER;
            privilegedUser = false;
            break;
        case 2:
        case 18:
            mode = INTERRUPT_REQUEST;
            break;
        case 3:
        case 19:
            mode = SUPERVISOR;
            break;
        case 23:
            mode = ABORT;
            break;
        case 27:
            mode = UNDEFINED;
            break;
        case 31:
            mode = SYSTEM_OR_USER;
            privilegedUser = true;
            break;
        default:
            exit(PENDING_CODE);
    }
}

int Registers::getStepAmount() {
    return thumbMode ? HALFWORD_SIZE : WORD_SIZE;
}

int Registers::getPC(){
    return reg15;
}

void Registers::setPC(int imm){
    reg15 = imm + getStepAmount();
}

uint32_t Registers::getReg(int index){
    if (index < 13)
        return unbankedReg[index];
    if (index < 15){
        return bankedReg[getBankedIndex(index)];
    }
    if (index == 15)
        return reg15 + getStepAmount();
}

void Registers::setReg(char index, int data){
    if (index < 13)
        unbankedReg[index] = data;
    else if (index < 15)
        bankedReg[getBankedIndex(index)] = data;
    else if (index == 15)
        reg15 = data;
}

int Registers::getCPSR(){
    return currentStatusReg;
}

void Registers::setCPSR(int data){
    currentStatusReg = data;
    setControlBits();
}

void Registers::setFlags(char mask, int data){
    int changeMask = mask<<24;
    int protectMask = ~changeMask;
    currentStatusReg &= protectMask;
    currentStatusReg |= (data & changeMask);
}

int Registers::getSPSR(){
    return savedStatusReg[mode];
}

void Registers::setSPSR(int data){
    savedStatusReg[mode] = data;
}

void Registers::branch(int imm){
    if (thumbMode)
        reg15 += (imm+1)*HALFWORD_SIZE;
    else
        reg15 += (imm+1)*WORD_SIZE;
}

bool Registers::isThumbMode(){
    return thumbMode;
}

void Registers::exchange(int address){
    bool prevThumbMode = thumbMode;
    thumbMode=(address&1);
    if(thumbMode){
        reg15 = address & ~1;
        if (prevThumbMode)
            reg15 -= HALFWORD_SIZE;
        DEBUG_OUT << "\nPC.t = "<< reg15 <<endl;
    }
    else{
        reg15 = address & ~2;
        if (!prevThumbMode)
            reg15 -= WORD_SIZE;
        DEBUG_OUT << "\nPC.a = "<< reg15 <<endl;
    }
}

void Registers::step(){
    reg15 += getStepAmount();
}

void Registers::status(){
    DEBUG_OUT<<"\nMODE: "<<(thumbMode?"THUMB":"ARM")<<"("<<getMode()<<")"<<" \tPC: "<<getPC()<<endl;
    DEBUG_OUT<<"R0: "<<unbankedReg[0]<<"\tR1: "<<unbankedReg[1]<<"\tR2: "<< unbankedReg[2]<<"\tR3: "<< unbankedReg[3]<< endl;
    DEBUG_OUT<<"R4: "<<unbankedReg[4]<<"\tR5: "<<unbankedReg[5]<<"\tR6: "<< unbankedReg[6]<<"\tR7: "<< unbankedReg[7]<< endl;
    DEBUG_OUT<<"R8: "<<getReg(8)<<"\tR9: "<<getReg(9)<<"\tR10: "<< getReg(10)<<"\tR11: "<< getReg(11)<< endl;
    DEBUG_OUT<<"R12: "<<getReg(12)<<"\tR13(SP): "<<getReg(SP)<<"\tR14(LR): "<< getReg(LR)<<"\tR15(PC+"<<noshowbase<<getStepAmount()<<showbase<<"): "<< getReg(PC)<< endl;
    DEBUG_OUT<<"CPSR: "<<currentStatusReg<< "\tSPSR:"<<getSPSR()<< endl;
}

bool Registers::canExecute(int cond){
    switch (cond)
    {
    case EQ:
        return (currentStatusReg & Z) != 0;
    case NE:
        return (currentStatusReg & Z) == 0;
    case CC:
        return (currentStatusReg & C) == 0;
    case MI:
        return (currentStatusReg & N) != 0;
    case PL:
        return (currentStatusReg & N) == 0;
    case VS:
        return (currentStatusReg & V) != 0;
    case GE:
        return canExecute(MI) == canExecute(VS);
    case LT:
        return canExecute(MI) != canExecute(VS);
    case ALWAYS:
        if(thumbMode) exit(FAILED_TO_DECODE);
        return true;
    default:
        cout << "Cond Undefined: " << Condition(cond).toString() << endl;
        exit(FAILED_TO_EXECUTE);
    }
    // if (cond == CS) return (currentStatusReg & 0x20_00_00_00) != 0;
    // else if (cond == VC) return (currentStatusReg & 0x10_00_00_00) == 0;
    // else if (cond == HI) return canExecute(CS) && canExecute(NE);
    // else if (cond == LS) return canExecute(CC) || canExecute(EQ);
    // else if (cond == GT) return canExecute(NE) && canExecute(GE);
    // else if (cond == LE) return canExecute(EQ) || canExecute(LT);
    return false;
}

string Condition::toString(){
    switch (value)
    {
    case EQ:
        return "{eq}";
    case NE:
        return "{ne}";
    case CC:
        return "{cc}";
    case PL:
        return "{pl}";
    case GE:
        return "{ge}";
    case LT:
        return "{lt}";
    case ALWAYS:
        return "";
    default:
        cout << "COND = " << value << endl;
        exit(FAILED_DECODED_TO_STRING);
    }
    return "Never";
}
